src/elements/common/logger/Logger.js

```diff
@@ -0,0 +1,200 @@
+function generateRandom(c: string): string {
+    /* eslint-disable no-bitwise */
+    const r = (Math.random() * 16) | 0;
+    const v = c === 'x' ? r : (r & 0x3) | 0x8;
+    return v.toString(16);
+    /* eslint-enable no-bitwise */
+}
+
+/**
+ * Generates a GUID/UUID compliant with RFC4122 version 4.
+ *
+ * @return {string} A 36 character uuid
+ */
+function uuidv4() {
+    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, generateRandom);
+}
+
+const SESSION_ID = uuidv4();
+const uniqueEvents: Set<string> = new Set();
+
+class Logger extends React.Component<Props> {
+    static defaultProps = {
+        onMetric: noop,
+    };
+
+    get uniqueEvents(): Set<string> {
+        return uniqueEvents;
+    }
+
+    get sessionId(): string {
+        return SESSION_ID;
+    }
+
+    /**
+     * Creates an event name meant for use with an event which meant to be logged once per file id
+     *
+     * @param {string} name - The event name
+     * @param {string} fileId - the file id
+     * @returns {string} A string containing the component and event name
+     */
+    createFileEventNameWith(name: string, fileId: string): string {
+        return `${fileId}::${this.createEventName(name)}`;
+    }
+
+    /**
+     * Creates an event name meant for use with an event which is unique and meant to be logged only once
+     *
+     * @param {string} name - The event name
+     * @returns {string} A string containing the component and event name
+     */
+    createEventName(name: string): string {
+        return `${this.props.source}::${name}`;
+    }
+
+    /**
+     * Checks to see if the specified event for the component has already been fired.
+     *
+     * @param {string} component - the component name
+     * @param {string} name - the event name
+     * @returns {boolean} True if the event has already been fired
+     */
+    hasLoggedEvent(name: string): boolean {
+        return this.uniqueEvents.has(name);
+    }
+
+    /**
+     * Invokes the provided metric logging callback.
+     *
+     * @param {string} type - the type of the event
+     * @param {Object} data  - the event data
+     * @param {string} name - the name of the event
+     */
+    logMetric(type: MetricTypes, name: string, data: Object): void {
+        const metric: ElementsMetric = {
+            ...data,
+            component: this.props.source,
+            name,
+            timestamp: this.getTimestamp(),
+            sessionId: this.sessionId,
+            type,
+        };
+
+        this.props.onMetric(metric);
+    }
+
+    /**
+     * Logs a unique metric event. Prevents duplicate events from being logged in the session.
+     *
+     * @param {string} type - the type of the event
+     * @param {Object} data  - the event data
+     * @param {string} name - the name of the event
+     */
+    logUniqueMetric(type: MetricTypes, name: string, data: Object): void {
+        let eventName;
+        if (type !== METRIC_TYPE_ELEMENTS_LOAD_METRIC) {
```

Comment:
Should this instead be referred to as a JS load metric? We may want to refer to events with API load times as load metrics (or not, just a thought)

Reply:
The event name can be used for this purpose